### DISCUSSION TITLE
Pytest doesn't *seem* to be using Python version installed with asdf

### DISCUSSION THREAD

**Discussion by audiodude** [pytest-dev/pytest#11605] [Q&A] [answered]
Cross posted from https://stackoverflow.com/questions/77463416/with-asdf-installed-python-why-does-pytest-use-a-differentpython-version-than-t

I have installed several Python versions using [asdf](asdf-vm.com). I am also using Pipenv to manage dependencies. Although I have Python 3.9.4 referenced in my Pipenv file, and asdf reports that I'm using Python 3.9.4, pytest reports that it's using 3.11.6.

For example:

```
tmoney@tmoney-linux:~/code/wp1$ asdf list python
  3.12.0
  3.7.16
 *3.9.4
tmoney@tmoney-linux:~/code/wp1$ cat .tool-versions 
python 3.9.4
tmoney@tmoney-linux:~/code/wp1$ cat Pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
fakeredis = "==2.4.0"
flask = "==2.3.2"
flask-session2 = "==1.3.1"
flask-cors = "*"
flask-gzip = "*"
gunicorn = "*"
kiwixstorage = "==0.8"
mwclient = "==0.9.3"
mwoauth = "==0.3.8"
mwparserfromhell = "==0.5.2"
oauthlib = "==2.1.0"
pyjwt = "==2.4.0"
pymysql = "==0.9.3"
pyparsing = "==3.0.9"
pytz = "==2022.7.1"
pyyaml = "==6.0.1"
redis = "==4.4.4"
requests = "==2.31.0"
rdflib = "~=6.0"
rq = "==1.13.0"
rq-scheduler = "==0.13.0"
validators = "*"
attrs = "==18.2.0"

[dev-packages]
pre-commit = "==2.21.0"
pytest = "==7.4.2"
pytest-cov = "==4.1.0"
yapf = "==0.28.0"
yoyo-migrations = "==7.3.2"

[requires]
python_version = "3.9.4"
tmoney@tmoney-linux:~/code/wp1$ python --version
Python 3.9.4
tmoney@tmoney-linux:~/code/wp1$ python3 --version
Python 3.9.4
tmoney@tmoney-linux:~/code/wp1$ asdf reshim python
tmoney@tmoney-linux:~/code/wp1$ pipenv run pytest
============================================ test session starts =============================================
platform linux -- Python 3.11.6, pytest-7.4.2, pluggy-1.3.0
<SNIP>
```

**Comment by audiodude** [comment, marked as answer]
Turns out I have to explicitly set the Python version on pipenv when using asdf:

```
pipenv --rm
pipenv --python 3.9.4
```

**Comment by blagh** [comment]
For me, also using `asdf`, part of the issue was that the python pinned in `.tool-versions` was different from `project.toml`. I had some dependencies installed from another project so those were working, but `pytest` for whatever reason wasn't.

Also for whatever reason, Poetry was still stuck on the other Python version until I updated to Poetry to 2.2.1 (perhaps I should have tried more cache clearing).
